### ISSUE TITLE
bug: cannot format typescript files due to "Error detected while processing BufWritePre Autocommands for "*" "

### ISSUE THREAD

**Issue by guoliang** [stevearc/conform.nvim#520]
### Neovim version (nvim -v)

nvim 0.10.1

### Operating system/version

MacOS 14.6.1

### Read debugging tips

- [X] I have read through the [debugging tips](https://github.com/stevearc/conform.nvim/blob/master/doc/debugging.md).

### Add the debug logs

- [X] I have set `log_level = vim.log.levels.DEBUG` and pasted the log contents below.

### Log file

```bash
Error detected while processing BufWritePre Autocommands for "*":
Error executing lua callback: ...Cellar/neovim/0.10.1/share/nvim/runtime/lua/vim/glob.lua:79: subcapture nesting too deep
stack traceback:
        [C]: in function 'match'
        ...Cellar/neovim/0.10.1/share/nvim/runtime/lua/vim/glob.lua:79: in function 'to_lpeg'
        ...eovim/0.10.1/share/nvim/runtime/lua/vim/lsp/_dynamic.lua:105: in function 'match'
        ...eovim/0.10.1/share/nvim/runtime/lua/vim/lsp/_dynamic.lua:73: in function 'get'
        ...eovim/0.10.1/share/nvim/runtime/lua/vim/lsp/_dynamic.lua:83: in function 'supports_method'
        .../Cellar/neovim/0.10.1/share/nvim/runtime/lua/vim/lsp.lua:779: in function 'get_clients'
        .../share/nvim/lazy/conform.nvim/lua/conform/lsp_format.lua:49: in function 'get_format_clients'
        ....local/share/nvim/lazy/conform.nvim/lua/conform/init.lua:400: in function 'has_lsp_formatter'
        ....local/share/nvim/lazy/conform.nvim/lua/conform/init.lua:458: in function 'format'

```

### Describe the bug

Exception thrown, when saving angular/typescript files.

Exception still throws even though I have commented out formatters

```lua
    formatters_by_ft = {
      lua = { 'stylua' },
      css = { 'prettierd', 'prettier', stop_after_first = true },
      less = { 'prettierd', 'prettier', stop_after_first = true },
      html = { 'prettierd', 'prettier', stop_after_first = true },
      -- javascript = { 'prettierd', 'prettier', stop_after_first = true },
      -- javascriptreact = { 'prettierd', 'prettier', stop_after_first = true },
      json = { 'prettierd', 'prettier', stop_after_first = true },
      markdown = { 'prettierd', 'prettier', stop_after_first = true },
      -- typescript = { 'prettierd', 'prettier', stop_after_first = true },
      -- typescriptreact = { 'prettierd', 'prettier', stop_after_first = true },
      yaml = { 'prettierd', 'prettier', stop_after_first = true },
    },

```

### What is the severity of this bug?

blocking (cannot use plugin)

### Steps To Reproduce

1. with Mason install prettierd, and prettier
2. create angular project
3. save any typescript file
4. see error

### Expected Behavior

On save, it should auto format

### Minimal init.lua

```Lua
-- DO NOT change the paths and don't remove the colorscheme
local root = vim.fn.fnamemodify("./.repro", ":p")

-- set stdpaths to use .repro
for _, name in ipairs({ "config", "data", "state", "cache" }) do
  vim.env[("XDG_%s_HOME"):format(name:upper())] = root .. "/" .. name
end

-- bootstrap lazy
local lazypath = root .. "/plugins/lazy.nvim"
if not vim.loop.fs_stat(lazypath) then
  vim.fn.system({
    "git",
    "clone",
    "--filter=blob:none",
    "--single-branch",
    "https://github.com/folke/lazy.nvim.git",
    lazypath,
  })
end
vim.opt.runtimepath:prepend(lazypath)

-- install plugins
local plugins = {
  "folke/tokyonight.nvim",
  {
    "stevearc/conform.nvim",
    config = function()
      require("conform").setup({
        log_level = vim.log.levels.DEBUG,
        -- add your config here
      })
    end,
  },
  -- add any other plugins here
}
require("lazy").setup(plugins, {
  root = root .. "/plugins",
})

vim.cmd.colorscheme("tokyonight")
-- add anything else here
```

### Additional context

Maybe similar to https://github.com/stevearc/conform.nvim/issues/393

I already tried to re-install the plugin, I also tried to re-install `prettier`, and `prettierd` none seems to help.

**Comment by thomashilzendegen** [comment]
> hmm, seems like removing the entire `mason` folder, and then reinstall the formatters (`prettier`) helped in my case.

Didn't work for me – I still have the error coming up...

**Comment by HeadCase** [comment]
I am also still getting this error, even after purging Mason. Does not appear constrained to prettier/typescript either.

**Comment by guoliang** [comment]
@thomashilzendegen  @HeadCase 

I assume you removed the mason folder located in `~/.local/share/nvim/mason`

The error came back again to me somehow, but I did the same procedure with removing `mason` folder, and it resolved my issue again.

**Comment by HeadCase** [comment]
I did, yes. I tried the same again to be doubly sure, and got the same error. I also completely blew away my nvim config (`~/.config/nvim` & `~/.cache/nvim` & `~/.local/state/nvim` & `~/.local/share/nvim`) all to no avail. It is worth noting that not only does the format-on-save autocommand not work, I cannot even get `:ConformInfo` to succeed. Running a standard `:checkhealth` reports:

```
conform: require("conform.health").check()

conform.nvim report
- Log file: /home/ubuntu/.local/state/nvim/conform.log
- ERROR Failed to run healthcheck for "conform" plugin. Exception:
  ....local/share/nvim/lazy/conform.nvim/lua/conform/util.lua:53: attempt to call field 'root' (a nil value)

```
Sadly, it would appear this means Conform dies before logging anything, as the log file does not exist: 
```
❯ ls ~/.local/state/nvim
 lazy   log   mason-registry-update   mason.log   sessions   shada   swap   undo

```
And for good measure, my `format.lua` configuration:
```
return {
  "stevearc/conform.nvim",
  log_level = vim.log.levels.DEBUG,
  opts = {
    formatters_by_ft = {
    json = { "prettier" },
    markdown = { "mdformat" },
    python = { "isort", "blue" },
    },
  },
}
```

**Comment by guoliang** [comment]
@HeadCase sorry to hear that removing the mason folder didn't help.

Maybe doesn't matter, but when I had the error I also had the same issue wherein `:ConformInfo` died before it showed me any debugging log.

Your plugin setup is smaller than mine. Did you try to comment out some of the languages to verify whether it is one of the formatter used which causes Conform to crash?

**Comment by stevearc** [comment]
The original error in the original bug report is a crash in the LSP logic, not in conform. It looks like it would happen if you call
```lua
vim.lsp.get_clients({bufnr = 0, method = "textDocument/formatting"})
```

If that is the case, your best bet is reporting this upstream. It looks like there's something weird with the filter pattern in the register options for the formatting method. This line might show you more data:
```lua
vim.lsp.get_clients({bufnr = 0})[1].capabilities["textDocument/formatting"]
```

**Comment by jonathan-elize** [comment]
So just adding onto this. I suddenly was getting the same issue. It seems like as @stevearc said it has to do with LSPs and not conform itself per say. What I did was disable the LSPs on my typescript files and reenable them once at a time to see what triggered the issue.

What I found is that my `eslint` lsp is what seemed to trigger the problem So on it's `on_attach` I disabled the formatting:

```typescriptreact
require("lspconfig")["eslint"]{
  on_attach = function(client, bufnr)
    client.server_capabilities.documentFormattingProvider = false -- 0.8 and later
    client.server_capabilities.documentRangeFormattingProvider = false -- 0.8 and later
  end,
}
```

**Comment by hchris21** [comment]
I have just noticed a new debugging approach and I tested it right away. The file that fails to be formatted by conform is formatted successfully when running prettier & prettierd via the [suggested approach here]("/path/to/file.txt").

Based on this, what could the next steps be @stevearc ?

**Comment by stevearc** [comment]
@hchris21 what that means is that the formatters are working properly but the crash in the LSP code is preventing us from running it. Try running `:=vim.lsp.get_clients({bufnr = 0, method = 'textDocument/formatting'})` I would expect that command to produce the same crash. If you run `:=vim.lsp.get_clients({bufnr = 0})`, you should be able to see which LSP clients are attached to the buffer, and thus might be causing the problem. The crash is happening [on this line](https://github.com/neovim/neovim/blob/f9bf64d74641f33961fe1d6c1592b65d2744298f/runtime/lua/vim/lsp/_dynamic.lua#L70), so presumably there is something wrong with the `documentSelector`. Try looking for the value of that to see if it's something reasonable (and thus there is a bug in how Neovim handles it) or if it's invalid (and thus there is a problem with the LSP server).

**Comment by hchris21** [comment]
Thank you @stevearc

You are right, running that command resulted in the same error, and after I checked the LSP clients, I have noticed that the only LSP with the 'textDocument/formatting' capability was `eslint`. Disabling it solved the formatting issue.

**Comment by Fasaloft** [comment]
Hi, maybe this will help. For me works this setup in LSP:

https://www.lazyvim.org/configuration/recipes#add-eslint-and-use-it-for-formatting

If I understand this correctly, there is some fight, who will format the TS files. This specifies only one formatter.

**Comment by evgkirov** [comment]
I had the same problem.

In my case, it was due to a file path that was too long. I had to reorganize my code a bit (was about to do that anyway).

Waiting for the proper fix in 0.11
https://github.com/neovim/neovim/issues/26520
